Lib/glyphsLib/builder/glyph.py

```diff
@@ -250,22 +250,20 @@ def effective_width(layer, glyph):
     # The width may be taken from another master via the customParameters
     # 'Link Metrics With Master' or 'Link Metrics With First Master'.
     font = glyph.parent
-    master = font.masters[layer.associatedMasterId or layer.layerId]
-    metrics_source = master.metricsSource
-    if metrics_source is None:
-        width = layer.width
-    else:
-        metric_layer = font.glyphs[glyph.name].layers[metrics_source.id]
-        if metric_layer:
-            width = metric_layer.width
-            if layer.width != width:
-                logger.debug(
-                    f"{layer.parent.name}: Applying width from master "
-                    f"'{metrics_source.id}': {layer.width} -> {width}"
-                )
-        else:
-            width = None
-    return width
+    master = font.masters[layer.layerId]
```

Comment:
This line is the important change of the patch, it says to only consider the master of this layer (if any) for the metrics source, and never the "associated master". For intermediate layers it means that there's no metrics source, whereas previously they used the metrics source of their parent layer, which was wrong.